### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by ZER0**

### CODE REVIEW

File: .travis.yml
Comment (CPerezz) [inline .travis.yml:16]: We should also publish the docs. But it's fine atm

File: Cargo.toml
Comment (CPerezz) [inline Cargo.toml:10]: Should be `tag = v0.6.1`

File: src/lib.rs
Comment (CPerezz) [inline src/lib.rs:18]: Too verbose.

Use:
```rust
use dusk_plonk::prelude::*;
```

File: src/macros.rs
Comment (CPerezz) [inline src/macros.rs:1]: A small description comment would be nice.

File: src/note.rs
Comment (CPerezz) [inline src/note.rs:19]: If that's refering to the `PoseidonCipher` I think the constant should be defined in `Poseidon252` repo. Not here.

File: src/note.rs
Comment (CPerezz) [inline src/note.rs:55]: Why not use `PoseidonCipher` here? See: https://github.com/dusk-network/Poseidon252/blob/17329418740028b571752e936b190a1f15f94c5a/src/cipher/cipher.rs#L84

File: src/note.rs
Comment (CPerezz) [inline src/note.rs:87]: We have it. So this should be removed or the comment should not be there.

File: src/note.rs
Comment (CPerezz) [inline src/note.rs:176]: If that's the `PoseidonCipher` nonce. The type is not correct. It should be `BlsScalar` as per https://github.com/dusk-network/Poseidon252/blob/17329418740028b571752e936b190a1f15f94c5a/src/cipher/cipher.rs#L110

File: src/note.rs
Comment (CPerezz) [inline src/note.rs:187]: No need to do this conversions. 
```rust
let value_commitment = (GENERATOR_EXTENDED * value_commitment)
                + (GENERATOR_NUMS_EXTENDED * blinding_factor);
```

File: src/note.rs
Comment (CPerezz) [inline src/note.rs:248]: We should simply return the error IMO. If the decryption fails you don't have anything to do with the Note.

File: src/note.rs
Comment (CPerezz) [inline src/note.rs:296]: Will this get the Index returned by the storage call from kelvin and store it?
If that's the case, this looks correct to me. Otherways doesn't since you can't decide where kelvin stores.

File: src/note.rs
Comment (CPerezz) [inline src/note.rs:325]: Shouldn't we fail here?

File: src/note.rs
Comment (CPerezz) [inline src/note.rs:393]: Shouldn't we add also the `Into<StorageScalar>` trait impl here?

File: .travis.yml
Comment (ZER0) [reply .travis.yml:16]: I copied one of the `.travis.yml` we have in our repo, what would a more recent one?
Or, should we already move out from `travis` also here and move to github actions? (In that case, I will file a follow up issue)

File: src/note.rs
Comment (ZER0) [reply src/note.rs:55]: First of all, because when this was implemented the first time we didn't have `PoseidonChiper` yet – that also why the `ECRYPTED_DATA_SIZE` is defined here. But once the `PoseidonChiper` was implemented, it was because the internals. It would be better if `PoseidonCipher` would use as internals bytes instead of `BlsScalar` – if we want to have this representation on contracts as well – but before changing either poseidon or phoenix-core it's better to cross that bridge and see what's the best approach.

File: src/note.rs
Comment (ZER0) [reply src/note.rs:87]: No, we don't.
Basically we need to remove that once we have the "proper" encrypted data type (see comment above) and we don't have it yet – since it would be better waiting, otherwise we've 50% chance to re-write the wrong part of the code.

File: src/note.rs
Comment (ZER0) [reply src/note.rs:176]: `phoenix-core` doesn't deal with `BlsScalar` directly (in the type structures I mean) but with JubJub's instead. The fact that Poseidon only deals with `BlsScalar` is an implementation details – meaning: from the point of the phoenix-core's consumer, Poseidon is a black box.

Having Nonce as BlsScalar in a public struct is an anomaly in the design, and leaking an internal type.

I talked with Togh and he doesn't care if it's JubJub or BlsScalar, and he's going to update the docs.

It's like the `value`, basically, that is passed and stored as u64 but then is converted internally as `BlsScalar: the consumers doesn't need to know that, they shouldn't care about that.

The alternative would be have the Nonce as simple sequence of 32 bytes (that "represent" a BlsScalar), from a design point of view is also valid.

File: src/note.rs
Comment (CPerezz) [reply src/note.rs:55]: @khovratovich @vlopes11 any suggestions?

File: src/note.rs
Comment (CPerezz) [reply src/note.rs:87]: kk. Fair enough

File: src/note.rs
Comment (CPerezz) [reply src/note.rs:176]: No, I'm happy then to have it as a jubjub or bls Scalar.

File: .travis.yml
Comment (CPerezz) [reply .travis.yml:16]: I don't have any problem staying with travis. Specially because the builds & tests here should not take too long since there's not any ZK stuff going on.

The block you're missing for the docs upload is the following:
```yaml
# Upload docs
  - |
      if [[ "$TRAVIS_OS_NAME" == "linux" && "$TRAVIS_RUST_VERSION" == "nightly" && "$TRAVIS_PULL_REQUEST" = "false" && "$TRAVIS_BRANCH" == "master" ]]; then
        cargo rustdoc --features nightly -- --html-in-header katex-header.html &&
        echo "<meta http-equiv=refresh content=0;url=dusk_plonk/index.html>" > target/doc/index.html &&
        git clone https://github.com/davisp/ghp-import.git &&
        ./ghp-import/ghp_import.py -n -p -f -m "Documentation upload" -r https://"$GH_TOKEN"@github.com/"$TRAVIS_REPO_SLUG.git" target/doc &&
        echo "Uploaded documentation"
      fi
```

i usually have it inside an `after_success:` block

Comment (CPerezz) [reply .travis.yml:16]: Replace `dusk_plonk` by the name of the crate and it should work without any problems.

File: Cargo.toml
Comment (vlopes11) [reply Cargo.toml:10]: Was it updated?

File: src/note.rs
Comment (vlopes11) [reply src/note.rs:55]: We have on the same structure a `JubJubScalar`, so in this case we do not have only bytes. Why does this apply?

File: src/note.rs
Comment (vlopes11) [inline src/note.rs:214]: If the operation is fallible then we should work with `Result`. Since the `Note` contains other scalars, then we have no strict requirement to use only bytes.

File: .travis.yml
Comment (ZER0) [reply .travis.yml:16]: I might just use the Makefile for that, in this case.

File: src/lib.rs
Comment (ZER0) [reply src/lib.rs:18]: Solving this because we discussed about when to use `prelude` and when not (basically for `dusk_plonk` prelude is necessary only when we're going to use the composer / plonk, not when we want to access only to the re-exported crates).

For future reference, we might change this approach in the future (but is not something we're going to do atm).

File: src/macros.rs
Comment (ZER0) [reply src/macros.rs:1]: I will add it, I'm also not sure if I want to keep it actually, since now (after the refactoring) is used way less.

File: src/note.rs
Comment (ZER0) [reply src/note.rs]: As discussed, it was needed because the hashing. Now we're using `to_hash_inputs()` and there is no need to such conversion here.

File: src/note.rs
Comment (ZER0) [reply src/note.rs:295]: Basically yes. This was used by the function `store_unspent_note` in the original code, and was set by the DB (Kelvin).

It's not used yet because that part is still missing (either `Rusk` that will take care of it or `phoenix-core` once is clear how the PoseidonTree are working from `Rusk-VM`).

File: src/note.rs
Comment (ZER0) [reply src/note.rs:324]: I added in the error refactoring mentioned previously also a failing here, since now we're returning a `Result`.

File: src/note.rs
Comment (ZER0) [reply src/note.rs:392]: I was thinking to add in a follow up PR, in order to land this – since the `StorageScalar` is also part of the `PoseidonTree` functionality that is still missing – but wen can also add the implementation directly in the Initial Commit.
You're the reviewer, let me know what's your preference.

Personally I'm more about progressive enhancement unless is a key function that prevent everyone to contribute to the repo.

File: src/note.rs
Comment (ZER0) [reply src/note.rs:212]: In order to use directly the `PoseidonCipher` as `encrypted_data`, that was my initial idea, we need few small changes in the `poseidon` repo, otherwise we might complicate a bit the code here.

Since `encrypted_data` is encapsulated and not exposed, I don't think it's a blocker for the landing. I will therefore proceed to merge this PR, open a PR for the `poseidon252` repo, and once the changes on the `poseidon252` are landed proceed to change the internals in the `Note`.

File: src/note.rs
Comment (CPerezz) [reply src/note.rs:390]: If you want to do it on a follw-ip PR, feel free to add the issue now.

**Review by CPerezz — APPROVED** [review]
LGTM_ACK

**Comment by ZER0** [comment]
As discussed I filed issue #8 (that is a blocker for issue #5) so we can merge the `phoenix-core`'s initial commit and unblock the works on circuits.

**Pull request merged**
